codex/sales/slotqueue.nim

```diff
@@ -0,0 +1,257 @@
+# slots.
+const DefaultMaxSize = 64'u
+
+proc `<`(a, b: SlotQueueItem): bool =
+  a.ask.pricePerSlot > b.ask.pricePerSlot or # profitability
+  a.ask.collateral < b.ask.collateral or     # collateral required
+  a.expiry > b.expiry or                     # expiry
+  a.ask.slotSize < b.ask.slotSize            # dataset size
+
+proc `==`(a, b: SlotQueueItem): bool =
+  a.requestId == b.requestId and
+  a.slotIndex == b.slotIndex
+
+proc new*(_: type SlotQueue,
+          maxWorkers = DefaultMaxWorkers,
+          maxSize = DefaultMaxSize): SlotQueue =
+
+  let mworkers = if maxWorkers == 0'u: DefaultMaxWorkers
+                 elif maxWorkers > maxSize: maxSize
+                 else: maxWorkers
+
+  SlotQueue(
+    # Add 1 to always allow for an extra item to be pushed onto the queue
+    # temporarily. After push (and sort), the bottom-most item will be deleted
+    queue: newAsyncHeapQueue[SlotQueueItem](maxSize.int + 1),
+    maxWorkers: mworkers,
+    workers: newSeqOfCap[SlotQueueWorker](mworkers),
+    running: false
+  )
+
+proc new*(_: type SlotQueueWorker): SlotQueueWorker =
+  SlotQueueWorker(
+    processing: newFuture[void]("slotqueue.worker.processing")
+  )
+
+proc init*(_: type SlotQueueItem,
+          requestId: RequestId,
+          slotIndex: uint64,
+          ask: StorageAsk,
+          expiry: UInt256): SlotQueueItem =
+
+  SlotQueueItem(
+    requestId: requestId,
+    slotIndex: slotIndex,
+    ask: ask,
+    expiry: expiry
+  )
+
+proc init*(_: type SlotQueueItem,
+           request: StorageRequest,
+           slotIndex: uint64): SlotQueueItem =
+
+  SlotQueueItem.init(request.id, slotIndex, request.ask, request.expiry)
+
+proc init*(_: type SlotQueueItem,
+          requestId: RequestId,
+          ask: StorageAsk,
+          expiry: UInt256): seq[SlotQueueItem] =
+
+  var i = 0'u64
+  proc initSlotQueueItem: SlotQueueItem =
+    let sqi = SlotQueueItem.init(requestId, i, ask, expiry)
+    inc i
+    return sqi
+
+  var sqis = newSeqWith[uint64](ask.slots, initSlotQueueItem())
+  Rng.instance.shuffle(sqis)
+  return sqis
+
+proc init*(_: type SlotQueueItem,
+          request: StorageRequest): seq[SlotQueueItem] =
+
+  return SlotQueueItem.init(request.id, request.ask, request.expiry)
+
+proc requestId*(self: SlotQueueItem): RequestId = self.requestId
+
+proc slotIndex*(self: SlotQueueItem): uint64 = self.slotIndex
+
+proc running*(self: SlotQueue): bool = self.running
+
+proc len*(self: SlotQueue): int = self.queue.len
+
+proc `$`*(self: SlotQueue): string = $self.queue
+
+proc `onProcessSlot=`*(self: SlotQueue, onProcessSlot: OnProcessSlot) =
+  self.onProcessSlot = some onProcessSlot
+
+proc activeWorkers*(self: SlotQueue): int = self.workers.len
+
+proc contains*(self: SlotQueue, sqi: SlotQueueItem): bool =
+  self.queue.contains(sqi)
+
+proc pop*(self: SlotQueue): Future[?!SlotQueueItem] {.async.} =
+  try:
+    let sqi = await self.queue.pop()
+    return success(sqi)
+  except CatchableError as err:
```

Comment:
I knew you'd ask :) I did have a good look into this previously (and again now), and the main issue is that `AsyncHeapQueue.pop` is catching all `CatchableErrors` and then re-raising them. It is catching any `CatchableError` that could occur awaiting the `getter` it created, which I believe could be limited to `CancelledErrors` only. However, it seems to me that this try/except is meant to be handling some other exception because the exception handling explicitly checks that the future is not cancelled, so it wouldn't make sense that only a `CancelledError` could be raised. The exception is also re-raised, which also doesn't make much sense if it's a `CancelledError`. @dryajov do you have more insight as to what errors could be caught here?